src/librustc_trans/adt.rs

```diff
@@ -162,284 +161,208 @@ pub fn represent_type<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
+                               substs: &Substs<'tcx>)->Struct<'tcx> {
+            let size_bytes = size.bytes();
+            let sized = from.sized;
+            let align = from.align.abi() as u32;
+            let packed = from.packed;
+            let fields;
+            if let Some(v) = variant {
+                fields = v.fields.iter().map(|field| {
+                    monomorphize::field_ty(tcx, substs, field)
+                }).collect::<Vec<_>>();
             }
-
-            if cases.len() == 1 && hint == attr::ReprAny {
-                // Equivalent to a struct/tuple/newtype.
-                return Univariant(mk_struct(cx, &cases[0].tys, false, t));
+            else {
+                fields = Vec::default();
             }
+            Struct{size: size_bytes, sized: sized, align: align, packed: packed, fields: fields}
+}
 
-            if cases.len() == 2 && hint == attr::ReprAny {
-                // Nullable pointer optimization
-                let mut discr = 0;
-                while discr < 2 {
-                    if cases[1 - discr].is_zerolen(cx, t) {
-                        let st = mk_struct(cx, &cases[discr].tys,
-                                           false, t);
-                        match cases[discr].find_ptr(cx) {
-                            Some(ref df) if df.len() == 1 && st.fields.len() == 1 => {
-                                return RawNullablePointer {
-                                    nndiscr: Disr::from(discr),
-                                    nnty: st.fields[0],
-                                    nullfields: cases[1 - discr].tys.clone()
-                                };
-                            }
-                            Some(mut discrfield) => {
-                                discrfield.push(0);
-                                discrfield.reverse();
-                                return StructWrappedNullablePointer {
-                                    nndiscr: Disr::from(discr),
-                                    nonnull: st,
-                                    discrfield: discrfield,
-                                    nullfields: cases[1 - discr].tys.clone()
-                                };
-                            }
-                            None => {}
-                        }
-                    }
-                    discr += 1;
-                }
-            }
+//This is used to determine the fields of the unsized variant of an enum.
+fn compute_null_fields<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
+                       adt: ty::AdtDef<'tcx>,
+                       substs: &Substs<'tcx>)->Vec<Ty<'tcx>> {
+    let cases = get_cases(cx.tcx(), adt, substs);
+    assert_eq!(cases.len(), 2); //More than two variants is not the null pointer optimization.
+    //We use unwrap heere because it is a bug if this function is called on something wherein the following fails.
+    let case = cases.iter().filter(|x| x.is_zerolen(cx)).next().unwrap();
```

Comment:
Shouldn't use `is_zerolen`, `nndiscr` is the index of the non-null variant, so the null variant has index `1 - nndiscr`.